usersession/client/client.go

```diff
@@ -336,33 +334,44 @@ type UserServiceUnitStatus struct {
 	NeedDaemonReload bool     `json:"needs-reload"`
 }
 
-func (client *Client) ServiceStatus(ctx context.Context, services []string) (map[int][]UserServiceUnitStatus, error) {
+func (client *Client) ServiceStatus(ctx context.Context, services []string) (map[int][]ServiceUnitStatus, map[int][]ServiceFailure, map[int]error) {
 	q := make(url.Values)
 	q.Add("services", strings.Join(services, ","))
+	errs := make(map[int]error)
 
 	responses, err := client.doMany(ctx, "GET", "/v1/service-status", q, nil, nil)
 	if err != nil {
-		return nil, err
+		errs[0] = err
+		return nil, nil, errs
 	}
 
-	info := make(map[int][]UserServiceUnitStatus)
+	stss := make(map[int][]ServiceUnitStatus)
+	failures := make(map[int][]ServiceFailure)
 	for _, resp := range responses {
-		if resp.err != nil {
-			if err == nil {
-				err = resp.err
+		// Parse status errors which were a result of failure to retrieve status of services
+		if agentErr, ok := resp.err.(*Error); ok && agentErr.Kind == "service-status" {
+			if errorValue, ok := agentErr.Value.(map[string]interface{}); ok {
+				if fs, err := decodeServiceErrors(resp.uid, errorValue, "status-errors"); err == nil && len(fs) > 0 {
+					failures[resp.uid] = append(failures[resp.uid], fs...)
+				}
 			}
 			continue
 		}
-		var si []UserServiceUnitStatus
-		if decodeErr := json.Unmarshal(resp.Result, &si); decodeErr != nil {
-			if err == nil {
-				err = decodeErr
-			}
+
+		// The response was an error, store the error
+		if resp.err != nil {
+			errs[resp.uid] = resp.err
 			continue
 		}
-		info[resp.uid] = si
+
+		var si []ServiceUnitStatus
+		if err := json.Unmarshal(resp.Result, &si); err != nil {
+			errs[resp.uid] = err
```

Comment:
for these kind of issues the other methods seem to save only the first?

Reply:
Changed this to only save the first